src/node/controller/messages.rs

```diff
@@ -225,6 +239,128 @@ impl Controller {
     async fn block_by_hash(&self, block_hash: &BlockHash) -> anyhow::Result<Option<Block>> {
         self.state.lock().await.get_block_by_hash(block_hash).await
     }
+
+    /// Actions to be performed to validate and store a state block
+    /// TODO: this assumes we will never get a live epoch block
+    async fn state_block_handler(&self, state_block: StateBlock) -> anyhow::Result<()> {
+        // TODO: here there should be a check for epoch blocks
+        if self.block_existed(&state_block.hash).await? {
+            info!("Block {} already exists!", state_block)
+        } else if state_block.verify_self_signature().is_err() {
+            info!("Block {} has invalid signature!", state_block)
+        } else {
+            self.process_valid_existing_state_block(state_block).await?
+        }
+        Ok(())
+    }
+
+    async fn process_valid_existing_state_block(
+        &self,
+        state_block: StateBlock,
+    ) -> anyhow::Result<()> {
+        match &state_block.previous {
+            Previous::Block(previous_hash) => {
+                // Either wants to send, receive or change
+                let maybe_previous_block = self.previous_as_account_info(previous_hash).await?;
+                if let Some(previous_state_block) = maybe_previous_block {
+                    self.process_block_with_previous(state_block, previous_state_block)
+                        .await?
+                } else {
+                    info!("Block before {} not found!", state_block)
+                }
+            }
+            Previous::Open => {
+                todo!("Received an open sub-block")
+            }
+        }
+        Ok(())
+    }
+
+    async fn process_block_with_previous(
+        &self,
+        mut state_block: StateBlock,
+        previous_state_block: StateBlock,
+    ) -> anyhow::Result<()> {
+        let is_send = state_block.balance < previous_state_block.balance;
+        let amount = if is_send {
+            previous_state_block
+                .balance
+                .checked_sub(&state_block.balance)
+        } else {
```

Comment:
There's also a equals condition too right? For changing reps I think.

Reply:
Yes, but this only decides if it's a send, otherwise it can be change or receive